staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go

```diff
@@ -307,19 +496,220 @@ func (q *delayingType) waitingLoop() {
+	// handle the different message actions
+	switch entry.action {
+	case waitForActionIsWaiting, waitForActionLenWaiting, waitForActionNextReady, WaitForActionWaitFill:
+		// all of these actions are processed in the call handleReportingAndSync()
+		if entry.returnCh == nil {
+			return
+		}
+		q.afterFill = append(q.afterFill, entry)
+		return
+	case waitForActionDoneWaiting:
+		// remove entry from the 'waiting' queue
+		existing, exists := knownEntries[entry.data]
+		if exists {
+			// remove the item from the 'waiting' queue
+			if q.logger != nil {
+				q.logger.Logf("waitingLoop: (X) entry forgotten: %v", entry.data)
+			}
+
+			heap.Remove(w, existing.index)
+			delete(knownEntries, entry.data)
+		}
+		return
+	case waitForActionQueue:
+		// add entry to either 'active' or 'waiting' queues (dependent on delay and settings)
+		if entry.returnCh != nil {
+			defer close(entry.returnCh)
+		}
+
+		// If the embedded workqueue is an EnhancedInterface we can access the de-duplication feature.
+		// Perform de-duplication between the 'waiting' queue against items in the 'active' queue unless
+		// disabled by PermitActiveAndWaiting.
+		if ei, ok := q.Interface.(EnhancedInterface); ok {
+			if !entry.options.PermitActiveAndWaiting && ei.IsQueued(entry.data) {
+				if q.logger != nil {
+					q.logger.Logf("waitingLoop: (X) PermitActiveAndWaiting dropping already active: %v", entry.data)
+				}
+				return
+			}
+		}
+
+		// drop if the incoming item if the caller wishes to keep the existing item if it exists.
+		existing, exists := knownEntries[entry.data]
+		if exists && entry.options.WhenWaiting == TakeExisting {
+			if q.logger != nil {
+				q.logger.Logf("waitingLoop: (X) TakeExisting dropping already waiting: %v", entry.data)
+			}
+			return
+		}
+
+		// delayed add (to 'waiting' queue)
+		if entry.readyAt.After(q.clock.Now()) {
+			// inform the metrics of the retry (moved from AddAfter())
+			q.metrics.retry()
+
+			// insert the entry into the 'waiting' queue with appropriate insertion strategy defined by options.Waiting
+			if !insert(w, knownEntries, entry) {
+				if q.logger != nil {
+					q.logger.Logf("waitingLoop: (X) insert dropped %v", entry)
+				}
+				return
+			}
+			if q.logger != nil {
+				q.logger.Logf("waitingLoop: --> waiting %v", entry)
+			}
+			return
+		}
+
+		// immediate add (to 'active' queue)
+		if exists {
+			// items in the 'waiting' queue must be longer than an immediate add, if we TakeLonger we must drop this add.
+			if entry.options.WhenWaiting == TakeLonger {
+				if q.logger != nil {
+					q.logger.Logf("waitingLoop: (X) Immediate+TakeLonger drops: %v", entry.data)
+				}
+				return
+			}
+			// This covers both the TakeShorter and TakeIncoming cases because this is incoming and must be the shortest
+			// right now too.  We will remove/pre-empt the item in the 'waiting' queue unless disabled.
+			if !entry.options.PermitActiveAndWaiting {
```

Comment:
No, this is within the implementation of delayingType so it will always be of the EnhancedInterface.  You only need to do that cast test when your object wants to contain a "DelayingInterface" or "EnhancedDelayingInterface".  That is why we only have it implemented to test for the "EnhancedInterface" which is because out delayingType contains an Interface and not an EnhancedInterface:

// delayingType wraps an Interface and provides delayed re-enqueuing
type delayingType struct {
	Interface

This is because we still have the function NewDelayingQueueWithCustomQueue() which takes an Interface based queue, so we can't always guarantee that the 'active' queue is our own implementation. 

It is worth noting that the RateLimitingQueue embeds a "DelayingInterface" so if we wanted to make any use of any of the enhanced functionality of the delaying queue we would have the perform the cast test for the delaying queue as we have done for the workqueue here.

Reply:
I see, I analyzed it wrongly. Thanks for the explanation. But please see my other comments wrt this topic.

> This is because we still have the function NewDelayingQueueWithCustomQueue() which takes an Interface based queue, so we can't always guarantee that the 'active' queue is our own implementation.

IMO we could change DelayingQueueConfig to demand a presence of a queue that always implements `IsQueued`. People that are updating their client-go will have to go over their codebase and update it. I am no expert, but shouln't such a breaking change be okay? I think we did something similar when demanding that users pass a context to a public functions.